### DISCUSSION TITLE
Support using prql as query language

### DISCUSSION THREAD

**Discussion by waynexia** [GreptimeTeam/greptimedb#3748] [Ideas]
### What problem does the new feature solve?

Add a new query language support for prql https://prql-lang.org, which allows writing queries in pipeline-like style. I tried a few queries in the [playground](https://prql-lang.org/playground/), one impressive example is querying *"the top 10 cities average billing info among the latest 100 records"*. It needs a CTE first to get the latest 100 records, and then calculate their average. But this CTE is hidden in prql:

```prql
from invoices
sort [-invoice_date]
take 100
group [billing_city] (
  aggregate [
    avg = average total
  ]
)
sort [-_frame.avg]
take 10
```

Generated SQL:
```sql
WITH table_1 AS (
  SELECT
    billing_city,
    total
  FROM
    invoices
  LIMIT
    100
)
SELECT
  billing_city,
  AVG(total) AS avg
FROM
  table_1
GROUP BY
  billing_city
ORDER BY
  avg DESC
LIMIT
  10
```

It also provides some short-hand time/date/interval annotations, we could also benefit from it.

### What does the feature do?

Support prql language. It would not have many things to do as it provides an embedded compiler library that can compile prql to SQL. We only need to touch the parser IMO.

**Comment by WenyXu** [comment]
It seems to suit me, Can I take this?

**Comment by waynexia** [comment]
> It seems like suit for me, Can I take this?

Thanks for your enthusiasm :heart: This feature is still under discussion, and we wonder if it suits GreptimeDB. You could also comment to tell what you think about it

**Comment by killme2008** [comment]
This query language looks interesting. Maybe we can give it a shot in the Flow project. @discord9

**Comment by tisonkun** [comment]
One thing I'm afraid is that, as in https://github.com/GreptimeTeam/greptimedb/issues/3557 -

> Third-party DSL without a vast ecosystem might be short-lived. [Flux](https://www.influxdata.com/blog/the-plan-for-influxdb-3-0-open-source/#heading4) itself is an example.

... whether it's even a good direction.

**Comment by tisonkun** [comment]
Move to Ideas until we have any concrete plan.

**Comment by sunng87** [comment]
Pipelined SQL variants has been popular these days. According to https://cloud.google.com/bigquery/docs/pipe-syntax-guide

The main use-cases are:

- Incremental query building for human
- Unix user experience for log analyzing
